### DISCUSSION TITLE
How to prevent "Credentials provider is closed"

### DISCUSSION THREAD

**Discussion by afgarcia86** [aws/aws-sdk-kotlin#1296] [General]
I have my pod configured to use an `AWS_ROLE_ARN` and `AWS_WEB_IDENTITY_TOKEN_FILE` to authenticate. And its working, however it seems like if my logic is called concurrently I often get the error: "Credentials provider is closed"

Has anyone run into this and/or happen to know how I can work around it?

**Comment by lauzadis** [comment]
Can you share a minimal reproduction of your issue? Are you using the client after it's closed?

**Reply by afgarcia86** [reply]
After taking a closer look I see that the `.use` function closes the client after being used.

Is the best practice to create a client each time I am going to use it? Or create the client once and use it for the lifetime of my application?

**Reply by ianbotsf** [reply]
A service client maintains some running state such as cached credentials and network connections which are cheaper to reuse than to recreate. If your greater concern is latency then keep a long-running service client throughout your application's lifetime. If your greater concern is memory/network/thread overhead _and_ your use of the service client is sporadic then `use`/`close` your client in a smaller scope.

**Reply by afgarcia86** [reply]
Thanks for the clarification
